### DISCUSSION TITLE
How to set formatting options for jdtls?

### DISCUSSION THREAD

**Discussion by parbo** [joaotavora/eglot#1222] [Q&A]
I've tried a bunch of different ways i found in old issues/discussions to set the `java.format.settings.url` and `java.format.settings.profile` settings, but I can't get it to work. How do I do it?

**Comment by parbo** [comment]
Maybe it doesn't work because I use eglot-java, it passes its own options: https://github.com/yveszoundi/eglot-java/blob/main/eglot-java.el#L146

**Reply by yveszoundi** [reply]
@parbo , `eglot-java` can now merge [custom initialization options](https://github.com/yveszoundi/eglot-java#lsp-server-initialization-options) on top of defaults. 

I added an example with the Eclipse Google formatting style.

**Comment by joaotavora** [comment]
I didn't know such a large package existed, but I'd just like to say
it's not an official package, I can't recommend it.  As far as I know
everything needed to support LSP jdtls is in eglot.el.

Have you read:
https://joaotavora.github.io/eglot/#Advanced-server-configuration  ?

João

**Comment by parbo** [comment]
I guess all the stuff I found in forums was outdated. After reading that I tried doing this:

```
  (add-to-list 'eglot-server-programs  `(java-mode . ("jdtls" :initializationOptions
  						     (:settings
  						      (:java
   						       (:format (:enabled t :settings (:url ,(expand-file-name (locate-user-emacs-file "cache/eclipse-java-google-style.xml"))
   											    :profile "GoogleStyle"))))
   						      :extendedClientCapabilities (:classFileContentsSupport t)))))
```

It seems to get sent correctly looking at the eglot logs, but jdtls still formats the same way.

**Comment by joaotavora** [comment]
Many servers don't document exactly the form of :initializationOptions
they support, if any.  I suggest you contact the server developers to
know if and how :initializationOptions is supported.

You can also try command-line options too to configure a server
as suggested in "5.2 User-specific configuration" of the Eglot manual.

What to pass there?  Maybe
https://github.com/eclipse/eclipse.jdt.ls#running-from-the-command-line
holds clues.

João

**Reply by cayhorstmann** [reply]
I have the same problem as the OP and spent some time looking into what the server wants and what lsp-java mode sends, where formatting works as intended. 

Here is the documentation of the initialization options of jdtls: https://github.com/eclipse-jdtls/eclipse.jdt.ls/wiki/Running-the-JAVA-LS-server-from-the-command-line#initialize-request. For formatting, the options to send are, in my case, with a specific xml rule file and profile:

```
  "initializationOptions": {
    "settings": {
      "java": {
        "format": {
          "enabled": true,
          "settings": {
            "url": "/home/cay/bin/cay-eclipse.formatter.xml",
            "profile": "cay"
          }
        }
      }
    }

```

With lsp-java-mode, these settings work. The rules in the xml file are applied. When debugging the lsp-mode traffic, the above JSON is sent.

With eglot, I try:

```
(add-to-list 'eglot-server-programs
             `(java-mode . ("jdtls"
                            :initializationOptions
                            (:settings
                             (:java
                              (:format
                               (:enabled "true"
                                         :settings
                                (:url "file:///data/cay/bin/cay-eclipse.formatter.xml"
                                      :profile "cay"))))))))
```

The eglot events window shows:

```
(:jsonrpc "2.0" :id 1 :method "initialize" :params
          (:processId 1057251 :rootPath "/home/cay/books/cj13/" :rootUri "file:///data/cay/books/cj13" :initializationOptions
                      (:settings
                       (:java
                        (:format
                         (:enabled "true" :settings
                                   (:url "file:///data/cay/bin/cay-eclipse.formatter.xml" :profile "cay")))))
...
```

Yet formatting with `eglot-format-buffer` does not work. It picks an entirely different formatter.

I feel that this is not an issue of failing to send the required initialization options to the language server. What would be a good way to debug why lsp-java mode can do this and eglot can't?

**Reply by joaotavora** [reply]
> I feel that this is not an issue of failing to send the required initialization options to the language server.

That's a reasonable conclusion, from what I see there's nothing blatantly wrong.  But sometimes some servers are very quirky and need information to be passed in very specific ways.  For example I see two `:settings` objects in your JSON object.  Is that exactly how it should look?

> What would be a good way to debug why lsp-java mode can do this and eglot can't?

If lsp-java mode allows it, you could try to look at its logs to see exactly which JSON object(s) it sends instead of the one (or along the ones) that Eglot was sending.

**Reply by cayhorstmann** [reply]
Yes, it is two `settings`, as you can see from the TypeScript in the link that I provided. The first one is in `InitializationOptions`, the other in `FormatOption`. 

Here is the section that is related to formatting in the lsp log:

```
Params: {
  "processId": 2373213,
  "rootPath": "/tmp/hello",
  ...
  "initializationOptions": {
    "settings": {
      "java": {
        ...
        "format": {
          "enabled": true,
          "settings": {
            "url": "/home/cay/bin/cay-eclipse.formatter.xml",
            "profile": "cay"
          },
          "comments": {
            "enabled": true
          },
          "onType": {
            "enabled": true
          },
          "tabSize": 3,
          "insertSpaces": true
        },
        ...
      },
      ...
  },
  ...
}
```

I now set all of these in eglot and see this in the logs:

```
(:jsonrpc "2.0" :id 1 :method "initialize" :params
          (:processId 2403949 :rootPath "/tmp/hello/" :rootUri "file:///tmp/hello" :initializationOptions
                      (:settings
                       (:java
                        (:format
                         (:enabled t :comments
                                   (:enabled t)
                                   :onType
                                   (:enabled t)
                                   :tabSize 3 :insertSpaces t :settings
                                   (:url "file:///data/cay/bin/cay-eclipse.formatter.xml" :profile "cay")))))
                                   ...
```
Here are the logs for invoking the formatter. In lsp-mode:

```
[Trace - 07:12:49 ] Sending request 'textDocument/formatting - (33)'.
Params: {
  "textDocument": {
    "uri": "file:///tmp/hello/NotHelloWorld.java"
  },
  "options": {
    "tabSize": 3,
    "insertSpaces": true,
    "trimTrailingWhitespace": true,
    "insertFinalNewline": true,
    "trimFinalNewlines": true
  }
}
```
And in eglot:
```
(:jsonrpc "2.0" :id 3 :method "textDocument/formatting" :params
          (:textDocument
           (:uri "file:///tmp/hello/NotHelloWorld.java")
           :options
           (:tabSize 8 :insertSpaces t :insertFinalNewline t :trimFinalNewlines t)))
```

Still, running `eglot-format-buffer` invokes the default formatter. Never mind the tab size; the brace style is completely different.

**Reply by joaotavora** [reply]
Ah I see.  The problem is that those options you set before are not what influences `eglot-format` at all:

Eglot works in Emacs, and integrates with it.  In Emacs, there are standard variables such as `tab-width` , to control tab size, `indent-tabs-mode` to control whether to insert tabs instead of spaces, and so on.  In particular, the `options` for `textDocument/formatting` are calculated from these four variables:

```lisp
:options (list :tabSize tab-width
                      :insertSpaces (if indent-tabs-mode :json-false t)
                      :insertFinalNewline (if require-final-newline t :json-false)
                      :trimFinalNewlines (if delete-trailing-lines t :json-false))
```

So, set the variables `tab-width`, `indent-tabs-mode` ,`require-final-newline`  and `delete-trailing-lines` and try again.

**Reply by cayhorstmann** [reply]
Like I said, never mind the tab width. I know how to fix that later. The problem is the brace style. I want jdtls to use the particular formatting rules in that xml file, which is something that it is able to do via lsp-java. But not eglot.

**Reply by joaotavora** [reply]
From the information you've given me, the only discrepancy I see in LSP communication was caused by the 4 inconsistent Emacs variables I noted. From what I can detect in a comparative analysis and from the information you've given me everything else is the same.

If you'd like to create a proper complete bug report with all the elements for a reproduction of this issue, we might be able to help you further.  Please follow the guide at https://joaotavora.github.io/eglot/#Troubleshooting-Eglot

**Comment by zsxh** [comment]
I think this is related to https://github.com/joaotavora/eglot/discussions/1197#discussioncomment-5515587

eglot sends empty `didChangeConfiguration` to jdtls.

**Comment by joaotavora** [comment]
On Thu, May 11, 2023 at 12:53 PM zsxh ***@***.***> wrote:
>
> I think this is related to #1197 (reply in thread)
>
> eglot sends empty didChangeConfiguration to jdtls.

I don't see the relation, can you clarify?

Also didChangeConfiguration is for project-specific
server configuration.  Here there is no want for
project specific server configuration -- as far as I can
tell -- rather user-specific configuration

The manual also explains how to to do project-specific
server configuration.

João

**Comment by zsxh** [comment]
Yes, I confirm that it's related to `didChangeConfiguration`.

```emacs-lisp
(package-initialize)
(global-company-mode)
(yas-global-mode)
(require 'eglot)
(add-to-list
 'eglot-server-programs
 `(java-mode . ("jdtls"
                :initializationOptions
                (:settings
                 (:java
                  (:completion (:guessMethodArguments t)))
                 :extendedClientCapabilities (:classFileContentsSupport t)))))

```

```
/tmp/demo
├── Abc.java
└── .git
```
```java
import java.util.ArrayList;
import java.util.List;

/**
 * Abc
 */
public class Abc {

    public static void main(String[] args) {
	String txt = "txt";
	List<String> alist = new ArrayList<>();
        alist // type dot to trigger completion
  }
}
```
open Abc.java, M-x eglot, you can see logs like this:

```
...
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
	        (:settings #s(hash-table size 1 test eql rehash-size 1.5 rehash-threshold 0.8125 data
				                           ())))
```

alist trigger completion

![1](https://github.com/joaotavora/eglot/assets/10348978/4a5d5b69-8e15-41c7-bfb6-03666fa43ed3)

![2](https://github.com/joaotavora/eglot/assets/10348978/7792984b-f050-46f4-9272-b93ae629c53b)

When I prevent sending empty didChangeConfiguration, config works fine to me.

```emacs-lisp
;; NOTE: hack workspace/didChangeConfiguration
(define-advice eglot-signal-didChangeConfiguration (:override (server) override)
    "Send a `:workspace/didChangeConfiguration' signal to SERVER.
When called interactively, use the currently active server"
    (interactive (list (eglot--current-server-or-lose)))
    (when-let ((settings (eglot--workspace-configuration-plist server)))
      (jsonrpc-notify
       server :workspace/didChangeConfiguration
       (list :settings settings))))
```

![3](https://github.com/joaotavora/eglot/assets/10348978/8c02bdb8-d607-4e32-b9f2-366576ed2bac)

**Reply by skyler544** [reply]
Thank you, this somehow seems to enable the use of the formatting options as well. I tested it on `GNU Emacs 29.1.90 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-10-26` with this procedure:

1. Run `$ emacs -q`  
2. Evaluate following in the scratch buffer:  
```lisp
(require 'eglot)
(add-to-list 'eglot-server-programs
             '(java-mode . ("jdtls" :initializationOptions
                            (:settings
                             (:java
                              (:format
                               (:enabled "true"
                                         :settings
                                         (:url "path-to-your-style.xml"))))))))
```  
3. Open a java file that is not formatted according to your style file (notice the very long lines):  
```java
package org.comments.services;

import org.comments.proxies.CommentNotificationProxy;
import org.comments.repositories.CommentRepository;

public class CommentService {
    private final CommentRepository commentRepository;
    private final CommentNotificationProxy commentNotificationProxy;

    public CommentService(CommentRepository commentRepository, CommentNotificationProxy commentNotificationProxy) {
        this.commentRepository = commentRepository;
        this.commentNotificationProxy = commentNotificationProxy;
    }
}
```  
4. `M-x eglot`  
5. `M-x eglot-format-buffer`  
6. Nothing happens  
7. Go back to step 1 and start over, this time including the code provided by zsxh.  
8. In step 6 the buffer will now be formatted, including line length limits and so on.

**Reply by cayhorstmann** [reply]
Do you mean the line `(remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)`? With that line added, I too get the custom formatter to work. 

Is that a safe fix, or just something to pinpoint the issue?

**Reply by joaotavora** [reply]
> Do you mean the line (remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)? With that line added, I too get the custom formatter to work.

Aha!  So it means that jdtls forgets previous command line options when workspace-specific options are also given.  Please @cayhorstmann comment this in the Emacs issue you created.

**Comment by joaotavora** [comment]
On Thu, May 11, 2023 at 1:50 PM zsxh ***@***.***> wrote:
>
> Yes, I confirm that it's related to didChangeConfiguration.

No, there's absolutely no relation to the link comment
you posted.  It would be the same as saying my
washing machine's manual and Mozart's opera are related
because they're both written in German.

:initializationOptions is NOT the same as "workspace
configuration".  Please read the manual link I posted.

I just know  it is possible to configure servers in two
fundamentally different ways, project-specific and
user-specific, as described in the manual.  It shouldn't
-- in theory -- need any hacks.

If user-specific configuration that this OP is trying
to achieve is not supported by jdtls server and only
project-specific configuration is absolutely required
for the jdtls server in particular -- that's for the jdtls
devs to clarify.  If that's the case then, again, the manual has
information of how to do that.

João

**Comment by joaotavora** [comment]
> Aha! So it means that jdtls forgets previous command line options when workspace-specific options are also given.

@zsxh I'm sorry if that's what you meant by the relation between user-specific and project-specific configuration.  
If it's true, it's a server-specific relation, but indeed a relation.

**Comment by skyler544** [comment]
> Do you mean the line `(remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)`?

I didn't know about this fix; I meant all of this:
```lisp
  (define-advice eglot-signal-didChangeConfiguration (:override (server) override)
    "Send a `:workspace/didChangeConfiguration' signal to SERVER.
When called interactively, use the currently active server"
    (interactive (list (eglot--current-server-or-lose)))
    (when-let ((settings (eglot--workspace-configuration-plist server)))
      (jsonrpc-notify
       server :workspace/didChangeConfiguration
       (list :settings settings))))
```

I've had this enabled since I commented and it hasn't caused any issues. I just tried your suggestion of `(remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)` and it also seems to enable the custom formatting. I'll test it for a few days and see if anything weird happens.

_Caveat lector_ I haven't looked into any of this deeply enough to see why these configuration snippets would or wouldn't work, I was just looking through the issue tracker and these discussions trying to see if I could find hints on configuring the `jdtls` formatter settings and found this thread.

**Reply by joaotavora** [reply]
> (remove-hook 'eglot-connect-hook 'eglot-signal-didChangeConfiguration)

To be clear, what this does is prevent Eglot from automatically sending _project_-specific configuration to the server, _any_ server. 

In `jdtls` it seems to matter, as project-specific configuration overrides (i.e. erases/does not "deep-merge" with) user-specific configuration given in the `initializationOptions`, so the default empty project-specific configuration will erase anything passed
in `initializationOptions`.

So it could be that another fix for this issue is to leave all Elisp alone and pass formatter settings with project-specific configuration, as described in the manual https://joaotavora.github.io/eglot/#Project_002dspecific-configuration

**Reply by cayhorstmann** [reply]
Passing formatter settings with project-specific configurations seems unattractive for my use case. When I have a bigger project, I use an IDE. I use Emacs for smaller tasks where I do not want to set up a project. 

Nevertheless, I read through the documentation to understand the process. I set up a file `.dir-locals.el` with the contents

```
((nil
  . ((eglot-workspace-configuration
      . (:jdtls (:initializationOptions
                 (:settings
                  (:java
                   (:format
                    (:enabled t 
                              :settings
                              (:url "file:///data/cay/bin/cay-eclipse.formatter.xml"
                                    :profile "cay"))))))
                )))))
```

guessing the syntax from [the Python/Go example](https://joaotavora.github.io/eglot/#Examples)

The custom formatter was not invoked. The `*Warnings` buffer contains: 

```
⛔ Warning (eglot): Server tried to register unsupported capability `workspace/didChangeWorkspaceFolders'
```

**Reply by yveszoundi** [reply]
@cayhorstmann, I believe that `eglot-workspace-configuration` assumes that you pass the contents of `:settings`, instead of the `initializaionOptions->settings` hierarchy that you've got.

#### Example for a dir-locals file

You don't need `:profile "cay"` bit, the LSP server will complain in the logs.... ("eglot events buffer" for a given project).

```emacs-lisp
((java-mode
  . ((eglot-workspace-configuration
      .
      (:java
       (:format
        (:settings
         (:url "file:///data/cay/bin/cay-eclipse.formatter.xml")
         :enabled t)))
      ))
  ))

```

#### Notes
- If the configuration is picked up properly, you should see it in the "eglot events buffer", along with registration capability request for `workspace/didChangeWorkspaceFolders`.
- Example of an eglot events buffer name: `*EGLOT (cay-secret-project-name/(java-mode)) events*`

**Reply by cayhorstmann** [reply]
Hi,  that syntax worked for me, thanks! I did need the `:profile "cay"` because my formatter has multiple profiles. I also managed to add in runtime configurations:

```
((java-mode
  . ((eglot-workspace-configuration
      .
      (:java
       (:format
        (:enabled t 
                  :settings
                  (:url "file:///data/cay/bin/cay-eclipse.formatter.xml"
                        :profile "cay")))
        :configuration
        (:runtimes
         [(:name "JavaSE-11" :path "/data/apps/jdk-11")
          (:name "JavaSE-17" :path "/data/apps/jdk-17")
          (:name "JavaSE-21" :path "/data/apps/jdk-21" :default t)]))
      ))
  ))
```

Now what would be the best way of achieving this globally, without a per-project file?

**Reply by joaotavora** [reply]
> Now what would be the best way of achieving this globally, without a per-project file?

It's in the manual.  Use `setq-default`.  But the right way is for jtdls to address the fact that empty workspace configuration completely overrides, instead of merging, `initializationOptions`.

Another possibly more correct approach is keep doing the `initializationOptions` (like [here](https://github.com/joaotavora/eglot/discussions/1222#discussioncomment-5871986) and add this:

```
(add-hook 'java-mode-hook (lambda ()
   (remove-hook 'eglot-connect-hook #'eglot-signal-didChangeConfiguration t))
```

So that -- in `java-mode` only -- Eglot refrains from sending workspace options.

**Reply by cayhorstmann** [reply]
I did that, and I added the runtime configurations like this:
```
(add-to-list 'eglot-server-programs
             `(java-mode . ("jdtls"
                            :initializationOptions
                            (:settings
                             (:java
                              (:format
                               (:enabled t 
                                         :settings
                                         (:url "file:///data/cay/bin/cay-eclipse.formatter.xml"
                                               :profile "cay")))
                              :configuration
                              (:runtimes
                               [(:name "JavaSE-11" :path "/data/apps/jdk-11")
                                (:name "JavaSE-17" :path "/data/apps/jdk-17")
                                (:name "JavaSE-21" :path "/data/apps/jdk-21" :default t)]))
                             ))
                         ))
```

The outcome: jdtls uses my custom formatter. jdtls does not use the default runtime. 

Why does act differently from using `.dir-locals.el`?

**Reply by joaotavora** [reply]
> I did that

What is "that"? I suggested two approaches in my post:  one to globally set `eglot-workspace-configuration`, another one to nullify it only in `java-mode`.

> Why does act differently from using .dir-locals.el?

`initializationOptions` is _not_ the same as `eglot-workspace-configuration`.  They're two different LSP interfaces, and the server does with them what it wants to do.  If you find this behavior incongruent, talk to the server devs.  

Here in the Eglot forum, we're mostly interested in things that Eglot _sends_ to the server (if it's sending them in the correct formats and at the correct time) or incorrect interpretations of things Eglot _receives_ from the server.

When I say "mostly", it doesn't mean we're _exclusively_ interested in that.  It's fine to chat about server things.  But, as much as it is easy to to regard Eglot as an abstraction umbrella over all things LSP in Emacs, it is _not_ meant to abstract away the servers.  LSP's stated goal is to do that, _not_ Eglot's. In Eglot -- perhaps contrary to `lsp-mode` -- the user must be minimally aware of what server they are running.  Some servers keep the needed brainpower to a comfortable minimum, some don't.  You picked one of the latter 😄  (though I understand you didn't have much choice).

**Reply by yveszoundi** [reply]
@cayhorstmann, I'm no Eclipse JDT LS expert, but you could try referencing the default runtime as `JAVA_HOME` inside the `:java` block. Add a new line prior the `(:format ....)` block, with the following contents:
`(:home "/data/apps/jdk-21")`

Not sure if the above works, but it might be worth a try. As @joaotavora mentioned, the Eclipse JDT LS server is a bit involved, and they might be able to answer better specific questions.

**Comment by HaraldKi** [comment]
As I went through all that trouble and learned a lot from this thread and others, let me try to summarize my solution:

1. Remove `eglot-signal-didChangeConfiguration` from `eglot-connect-hook` [as described above](https://github.com/joaotavora/eglot/discussions/1222#discussioncomment-8161713). Not doing so will end you up in the madhouse, as there is no way to understand why nothing works. I opened a [bug for eclipse.jdt.ls](https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3108#issuecomment-2018589030). 

2. These `initializationOptions` work for me.
```elisp
(add-to-list
   'eglot-server-programs
   `((java-mode hk-java-mode)
     .
     ("jdtls"
      :initializationOptions
      (:settings
       (:java
        (:format (:enabled t :profile "harald-variant" :settings (:url "/home/harald/.emacs.d/eclipse-formatter.xml"))
         :settings (:url "/home/harald/.emacs.d/eclipse-settings.prefs")
         ))))))
```
The details of how to structure the lispified JSON [are documented](https://github.com/eclipse-jdtls/eclipse.jdt.ls/wiki/Running-the-JAVA-LS-server-from-the-command-line#initialize-request), though it is no doubt a painful to piece it together from all those interfaces, not the least because a `:settings` field may appear in at least three (sub)interfaces.

For better control I also did this:

3. [Download](https://download.eclipse.org/jdtls/milestones/?d) a recent version of eclipse.jdt.ls. Don't get confused by the non-numerical sorting when picking the most recent one. 1.33 works for Java-21, [1.34 is for Java-22](https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3131#issuecomment-2046025372) which gets really important when [using preview features](https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3131).

4. Use a script to best control how jdtls is started, something like:
```bash
#!/bin/bash
# points to my current used version of jdt.ls
JDTLS=$(readlink -f ~/tools/Store/jdt-language-server)
VERSION=${JDTLS#*-server-}
# set it explicitly to be sure you know where this all ends up
DATA=~/work/jdtls-data
#DEBUG="-agentlib:jdwp=transport=dt_socket,address=127.0.0.1:8888,server=y,suspend=n"
#TRACE=-Djdt.ls.debug=true
cd "$DATA"
exec -a "jdtls($VERSION): java" java \
	-Declipse.application=org.eclipse.jdt.ls.core.id1 \
	-Dosgi.bundles.defaultStartLevel=4 \
	-Declipse.product=org.eclipse.jdt.ls.core.product \
	-Dlog.level=ALL \
        $DEBUG $TRACE \
	-Xmx1G \
	--add-modules=ALL-SYSTEM \
	--add-opens java.base/java.util=ALL-UNNAMED \
	--add-opens java.base/java.lang=ALL-UNNAMED \
	-jar "$JDTLS/plugins/org.eclipse.equinox.launcher_"*.jar \
	-configuration "$JDTLS/config_linux" \
	-data "$DATA"
```
The script assumes that you have set up the JDK you want to use in your `PATH`. In this case, setting `JAVA_HOME` should never be necessary. Rather check and make sure it is **not** set.

5. When changing the jdt.ls version or when things start to behave weird, stop the jdt.ls server(s) and clear your `$DATA` directory. I had very annoying behavior sometimes and clearing `$DATA` usually brought back some sanity.

6. Check `$DATA/.metadata/.log` for hints if things go wrong, though it does not contain a lot.

**Comment by jypma** [comment]
Thanks for the tips on here! `eglot-format` seems to respect this formatting fine, however any code actions (e.g. "generate toString()") still deliver code with tabs in them (my formatting settings specify spaces). I think I recall `lsp-mode` did manage to pull that off.

Is code generation working according to formatting settings for people trying these settings?

My eglot configuration is just this:
```elisp
(add-hook 'java-mode-hook (lambda ()
                            (remove-hook 'eglot-connect-hook #'eglot-signal-didChangeConfiguration t)))

(add-to-list 'eglot-server-programs
             '(java-mode . ("jdtls" :initializationOptions
                            (:settings
                             (:java
                              (:format
                               (:enabled "true"
                                         :settings
                                         (:url "/home/jan/.emacs.d/eclipse-format-jan.xml"))))))))
```

**Reply by HaraldKi** [reply]
This has only partially to do with the lsp-server. Yes, my jdtls sends a tab character, but I don't see it after the expansion because ultimately `yas--indent` is run which is somewhat controlled by the customizable variable `yas-indent-line`. I have never touched it and it was set to `auto`, which will run `indent-according-to-mode` which, I think, does the tab-to-space conversion. I am also using `java-mode` (as opposed to java-ts-mode) and my `C-h v indent-line-function` is `c-indent-line`. (Though I am not 100% sure this does the conversion.)

**Reply by huangfeiyu** [reply]
Hi @HaraldKi @parbo @cayhorstmann @jypma @joaotavora Just want to let you know, developers of jdtls are fixing the deviation. See https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/3108. This is great!

**Comment by cayhorstmann** [comment]
I think I finally made some progress. The key is not to use initialization options, but a default workspace configuration. Then the formatter works reliably. See https://horstmann.com/unblog/2025-09-20/ Of course, if JDTLS fixes that bug, then one can go back to initialization options. I don't know if there is an advantage of one over the other, if they both work.

**Reply by huangfeiyu** [reply]
Good blog! The maintainers of jdtls tried to fix the bug, actually they included the bug fix in one milestone release, but right after the release, many vscode users reported bugs, so they reversed the fix and said, he will try another time. But months after, no progress. Maybe, it's very hard to fix it.

**Reply by huangfeiyu** [reply]
@cayhorstmann The said issue has been closed by eclipse.jdtls.

**Reply by cayhorstmann** [reply]
@huangfeiyu Thanks for letting me know. Did you try the latest build? Does it now work with `initializationOptions`?

**Reply by huangfeiyu** [reply]
@cayhorstmann yes, I am using it. I use autobuid and importOrder, the two work correctly now, it didn’t work before.

**Reply by cayhorstmann** [reply]
Ok, thanks. It works for me too. I updated https://horstmann.com/unblog/2025-09-20

**Reply by joaotavora** [reply]
>  I updated https://horstmann.com/unblog/2025-09-20

Lol, have you?

> The Emacs LSP client is bare bones and sparsely documented. The Eclipse LSP server is buggy and sparsely documented. Their maintainers are not working together. 

dammn, not looking forward to my annual performance review with you, baws!

EDIT: this is an attempt at humor!  :)
